### PR TITLE
fix(vcs): overflow error when repo contains large number of files

### LINKED ISSUE #1162: GitHandler.getFiles() crashes on very long file list
### Current Behavior

If your project directory contains a very large list of files, _even if they are marked to be excluded_, the `GitHandler.getFiles()` method will crash.

### Expected behavior

A long list of files in a repo should not cause any problems.

### Reproducible example

I don't have a repo with it, but you can create a bare-bones project and install a huge number of large npm packages...

### Workaround

Sadly the only current workaround is to remove the large directories. In many cases it's okay because those directories will be installed in the build staging folder automatically, but it's not ideal, hence the `priority:high` tag.

### Suggested solution(s)

We need to revisit both how we list those files (ideally we don't have to list ignored files to begin with) and also to stream the result instead of buffering.

### Your environment

Reported on v0.10.8 and reproduced on latest master.

### PR DESCRIPTION
**Pull request by edvald**
**What this PR does / why we need it**:

We now run fewer git commands to list files in a directory, and avoid
having to list every file before later filtering out ignored files.
We also use streaming to avoid any buffering issues, and for improved
efficiency.

**Which issue(s) this PR fixes**:

Fixes #1162

### CODE REVIEW

**Comment by eysi09** [comment]
Does the change anything for #1097?

File: garden-service/src/vcs/git.ts [hunk 1]
```diff
@@ -74,60 +75,86 @@ export class GitHandler extends VcsHandler {
...
-
-        return [lsIgnoredFiles, lsUntrackedIgnoredFiles]
-      }))
-
-      ignored = uniq(flatten(await Bluebird.map(commands, async (cmd) => git(...cmd, path))))
-    } catch (err) {
-      // if we get 128 we're not in a repo root, so we get no files
-      if (err.exitCode !== 128) {
-        throw err
+    // List modified files, so that we can ensure we have the right hash for them later
```
Comment (eysi09) [inline garden-service/src/vcs/git.ts]: It might be useful to document the sequencing above the function since the control flow is a little hard to follow.

File: garden-service/src/vcs/git.ts [hunk 2]
```diff
@@ -74,60 +75,86 @@ export class GitHandler extends VcsHandler {
...
+    const trackedButIgnored = new Set(flatten(
+      await Promise.all(this.ignoreFiles.map(f =>
+        git("ls-files", "--ignored", "--exclude-per-directory", f),
+      )),
+    ))
+
+    // We run ls-files for each ignoreFile and do a manual set-intersection (by counting elements in an object)
+    // in order to optimize the flow.
+    const paths: { [path: string]: number } = {}
+    const output: VcsFile[] = []
```
Comment (eysi09) [inline garden-service/src/vcs/git.ts]: Rename to `files` or similar?

**Comment by edvald** [comment]
No, this doesn't address that issue.

File: garden-service/src/vcs/git.ts (hunk 1 above)
Comment (eysi09) [reply garden-service/src/vcs/git.ts]: Regardless, we should at least add a comment saying that this function returns files filtered against ignore files, includes and excludes, since the name is pretty generic.
